### ISSUE TITLE
TelemetryModules are not loaded when using TelemetryConfiguration.CreateDefault

### ISSUE THREAD

**Issue by xavierdecoster** [microsoft/ApplicationInsights-dotnet#1503]
For a work-around for the heartbeat initial interval bug https://github.com/microsoft/ApplicationInsights-dotnet/issues/1298, the recommendation was to use `TelemetryConfiguration.CreateDefault()` to ensure the `applicationinsights.config` file is also read before doing any additional configuration in code via the SDK. (see [this discussion](https://github.com/microsoft/ApplicationInsights-dotnet/issues/1298#issuecomment-558283325))

However, this doesn't load any `TelemetryModules` as configured in the `applicationinsights.config` file.

See this line of code passing in a `null` arg for the modules:
https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs#L369

Not sure whether issue https://github.com/microsoft/ApplicationInsights-dotnet/issues/1018 is similar, but it sure sounds related.

How can we work around this limitation? Is there a way to force load these from config, given the `TelemetryConfigurationFactory` is internal?

**Comment by xavierdecoster** [comment]
@cijothomas if there is any way to initialize the TelemetryModules ourselves as configured in the applicationinsights.config file, and attach it to our own TelemetryConfiguration (as opposed to TelemetryConfiguration.Active), I'd love to know. 

Basically:
* we call `TelemetryConfiguration.CreateDefault()` and attach our own DiagnosticsTelemetryModule to work-around #1298 
* we use this TelemetryConfiguration instance everywhere
* we no longer depend on TelemetryConfiguration.Active in our code
* this all works fine, except TelemetryModules are not loaded, even though configured in the config file

**Comment by cijothomas** [comment]
For all modules,
var mod = new module();
module.Initialize(config);
should work.

There is a way to load config from xml but not sure if it loads modules correctly.
https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs#L379

**Comment by xavierdecoster** [comment]
> There is a way to load config from xml but not sure if it loads modules correctly.
> https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs#L379

That's basically the other issue https://github.com/microsoft/ApplicationInsights-dotnet/issues/1018 I referenced, which also doesn't load the modules correctly (`null` argument on [this line](https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs#L387))

So by the sound of it, we should add code for each module we have configured in the config file as that section of the file is basically ignored at the moment, unless we use `TelemetryConfiguration.Active` which is exactly what we don't want.

**Comment by xavierdecoster** [comment]
The getter of `TelemetryModules.Instance` seems to query for the modules configured in the  `applicationinsights.config` file. Correct me if I'm wrong, but that's the effect I see.

This enables us to enumerate over each `TelemetryModule` configured in the `applicationinsights.config`, and properly initialize them against our custom `TelemetryConfiguration` object.

```
foreach (var telemetryModule in TelemetryModules.Instance.Modules)
{
    telemetryModule.Initialize(telemetryConfiguration);
}
```

However, as the API docs state [here](https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryModules.cs#L8), the `TelemetryModules` type is not meant to be used in our code... and I think the goal is to migrate away from the static infrastructure such as `TelemetryConfiguration.Active` and `TelemetryModules.Instance` and the like. 

For now, the above seems to unblock, though it'd be nice if these were initialized automatically when creating the `TelemetryConfiguration` taking into account the `applicationinsights.config` file (which happens for both `TelemetryConfiguration.CreateDefault` and `TelemetryConfiguration.CreateFromConfig` APIs). 

Fixing that seems like a prerequisite for being able to move off `TelemetryConfiguration.Active` completely without having to rely on `TelemetryModules.Instance`.

**Comment by xavierdecoster** [comment]
@cijothomas given a `TelemetryConfiguration` object, is there a way to query which `TelemetryModules` have been initialized against the instance? Would be useful to be able to harden against unexpected changes with test coverage, but in order to do so one needs to be able to query what has been initialized.

**Comment by cijothomas** [comment]
> @cijothomas given a `TelemetryConfiguration` object, is there a way to query which `TelemetryModules` have been initialized against the instance? Would be useful to be able to harden against unexpected changes with test coverage, but in order to do so one needs to be able to query what has been initialized.

There is no api to find this today.

**Comment by xavierdecoster** [comment]
> There is no api to find this today.

This could be useful to have so we can assert that certain modules have been initialized by our code. Any plans to add this functionality? Do you want me to report a separate issue?

**Comment by jezzsantos** [comment]
I am experiencing the same issue. 
The guidance clearly states that we should move away from `TelemetryConfiguration.Active` and instead use `TelemetryConfiguration.CreateDefault()` but the fact that the `<TelemetryModules>` node in the `ApplicationInsights.config` file are not loaded means that we cant move away from using `Active`.

I'm reporting that replacing `TelemetryConfiguration.Active` with `TelemetryConfiguration.CreateDefault()` means no more telemetry in AI in Azure.

Is there a fix for this, or progress on this?

**Comment by cijothomas** [comment]
> I am experiencing the same issue.
> The guidance clearly states that we should move away from `TelemetryConfiguration.Active` and instead use `TelemetryConfiguration.CreateDefault()` but the fact that the `<TelemetryModules>` node in the `ApplicationInsights.config` is being ignored means that we can't move away from using `Active`. If we continue to rely on `ApplicationInsights.config` (for .NET projects).
> 
> I'm reporting that by replacing `TelemetryConfiguration.Active` with `TelemetryConfiguration.CreateDefault()` means no more telemetry in AI in Azure. (in v 2.12.0)
> 
> Is there a fix for this, or progress on this?

If you are on .NET (not .NET Core), then TC.Active is valid and it'd read ai.config as well.

The overall issue about heartbeat, lack of easy way of configure it in .net core are still unsolved issues. No ETA as no one is actively working on this issue right now. Will update with ETA when someone from the team picks up this task.

**Comment by jezzsantos** [comment]
@cijothomas thanks for the clarification. 

p.s. Could I suggest, for others coming behind me, to make this clearer in the documentation?
Right now, it sounds like the directive for .NET (+ApplicationInsights.config) is to also go to`TC.CreateDefault()` which will break telemetry entirely.

**Comment by cijothomas** [comment]
@jezzsantos Help me once me - where do you see TC.CreateDefault() being recommended for .NET? From what I see in docs, that recommendation exist only for .net core.
Even in Visual Studio, Tc.Active shows obsolete warning only for .NET Core apps, not for .NET Framework apps.

**Comment by jezzsantos** [comment]
Hi @cijothomas,

It is not explicitly recommended for .NET.  I am asking for clarity and inclusion because of the omission, for those following behind.
